### ISSUE TITLE
Options for adding search

### ISSUE THREAD

**Issue by tijmenb** [alphagov/tech-docs-template#66]
I've been thinking how to add search. Have come up with 3 or so options.

## 1. Javascript

We could use [lunrjs](http://lunrjs.com/) to do search. It's a lightweight solr-clone in JS.

It would probably work by generating a JSON lunr database on each build. We'd create a frontend that uses lunr to show the search results.

I think this is what https://github.com/manastech/middleman-search is trying to do.

Downsides:

- It's a lot of javascript that needs to be executed during build phase. You'd need node installed or something?
- It's probably not accessible, doesn't work without Javascript

## 2. External service - push

On each build we create a JSON blob with all the content. This is then `POSTed` to the external service during deploy. This service indexes it, probably in elasticsearch.

If a user searches they'd be redirected to this search service, which serves results. We can have a sprinkle of JS on top that does auto-suggestion and such.

Downsides:

- This needs authentication, otherwise anybody can post

## 3. External service - pull

Same as above, but instead of _sending_ the JSON, the service scrapes the HTML of the site. Or pulls in the JSON.

After each deploy, the deploy script would only ping the search service to tell it to refresh. 

This looks like what https://www.algolia.com does (used by https://middlemanapp.com)

Downsides:

- Site may not be in sync with what's deployed, because there's lag between deploy & the indexer

Thoughts?

**Comment by 36degrees** [comment]
My preference would be definitely towards option one, as I'd like to keep this as self-contained and easy to deploy as possible. I think we should definitely at least explore it as an option and work out what its limitations are.

It looks like we could probably at least build the lunr index in Ruby – http://egardner.github.io/posts/2015/implementing-client-side-search-in-middleman/.

I guess the biggest question is whether search can be seen as an enhancement, and whether it's ok for it only to work for users with Javascript enabled.

**Comment by 36degrees** [comment]
Hah, wow, middleman-search actually invokes lunr.js with a V8 interpreter? Crazy kids.

**Comment by heathd** [comment]
@36degrees I had a play around with using `middleman-search` and ran into issues with a race condition when using ruby racer in combination with middleman's development mode. This caused middleman to hang intermittently. I worked around it using https://github.com/alphagov/middleman-search/commit/19df7578dc20621b60dfbc5e4dc986ac4f064c6b

with this setup I was able to build the lunrjs index and run a search client-side, however the styling and layout of the search interface needed work

I had assumed that the lunr index would be a more complex data structure than it actually is, so hadn't even considered building the index in ruby. But I think that's a much nicer approach.

I'm keen to use the tech docs format for GOV.UK Pay's team manual, and having search is a blocker to that, so we may get around to working on this at some point.
